x-pack/plugins/security_solution/public/common/utils/global_query_string/index.ts

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import * as H from 'history';
+import { parse, ParsedQuery, stringify } from 'query-string';
+import { useCallback, useEffect, useMemo } from 'react';
+
+import { url } from '@kbn/kibana-utils-plugin/public';
+import { isEmpty, pickBy } from 'lodash/fp';
+import { useHistory } from 'react-router-dom';
+import { useDispatch } from 'react-redux';
+import {
+  decodeRisonUrlState,
+  encodeRisonUrlState,
+  getParamFromQueryString,
+  getQueryStringFromLocation,
+} from '../../components/url_state/helpers';
+import { useShallowEqualSelector } from '../../hooks/use_selector';
+import { globalUrlParamActions, globalUrlParamSelectors } from '../../store/global_url_param';
+import { useRouteSpy } from '../route/use_route_spy';
+import { getLinkInfo } from '../../links';
+
+/**
+ * Adds urlParamKey and the initial value to redux store.
+ *
+ * Please call this hook at the highest possible level of the rendering tree.
+ * So it is only called when the application starts instead of on every page.
+ *
+ * @param urlParamKey Must not change.
+ * @param onInitialize Called once when initializing.
+ */
+export const useInitializeUrlParam = <State>(
+  urlParamKey: string,
+  /**
+   * @param state Decoded URL param value.
+   */
+  onInitialize: (state: State | null) => void
+) => {
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    const initialValue = getParamFromQueryString(
+      getQueryStringFromLocation(window.location.search),
```

Comment:
Might it be better to keep in sync with `react-router` by getting this from `useLocation()`?

Reply:
It is a hook now and I can call `useLocation`! Let me fix it.